### PR TITLE
[Pubsub] Using OBOD command batch for both reference counting and wait for object eviction

### PR DESCRIPTION
**Pull request by rkooo567**
## Why are these changes needed?

Using the command batch for ref count + object eviction. 

Note that the original proposal includes the command batch into the long polling request, but it will do that in a separate request (PubsubCommandBatch)

I also found the thread-safe issue while developing the command batch, so I made the subscriber thread-safe

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: src/ray/core_worker/core_worker.cc
Comment (rkooo567) [inline src/ray/core_worker/core_worker.cc:2448]: Unnecessary because the subscriber will cancel it from now on.

File: src/ray/core_worker/core_worker.cc
Comment (rkooo567) [inline src/ray/core_worker/core_worker.cc:2447]: We need to publish the message. Previously, it just replies the failure status, but this won't work for the command batch (command batch should always succeed)

File: src/ray/core_worker/core_worker.cc
Comment (rkooo567) [inline src/ray/core_worker/core_worker.cc:2482]: Any way to do it with switch? Please lmk

File: src/ray/core_worker/core_worker.cc
Comment (rkooo567) [inline src/ray/core_worker/core_worker.cc:2602]: Same as above

File: src/ray/core_worker/reference_count.cc
Comment (rkooo567) [inline src/ray/core_worker/reference_count.cc:800]: Moving lock to a subscriber is a TODO in the future PR.

Comment (rkooo567) [reply src/ray/core_worker/reference_count.cc:800]: Actually, I found an issue from thread safe, so I fixed it in this PR.

File: src/ray/core_worker/core_worker.cc
Comment (iycheng) [inline src/ray/core_worker/core_worker.cc:2465]: I know it's not your change, but we should update it to:

```
RAY_LOG(DEBUG) << "Reference for object " << object_id << " has already been freed.";
```

Before:
it'll do string formating no matter what

After:
it'll only have extra overhead when log==DEBUG

File: src/ray/core_worker/core_worker.cc
Comment (iycheng) [inline src/ray/core_worker/core_worker.cc:2478]: I'd prefer 
```
if () {
} else if{
} else {
LOG(FATAL) << "Unsupported type"
}
```

In this way, later if you add extra things, it can be added easily.

File: src/ray/core_worker/core_worker.cc
Comment (iycheng) [reply src/ray/core_worker/core_worker.cc:2482]: Check here https://developers.google.com/protocol-buffers/docs/reference/cpp-generated#oneof

I think `command. command_message_one_of_case()` should help

File: src/ray/core_worker/core_worker.cc
Comment (iycheng) [inline src/ray/core_worker/core_worker.cc:2489]: RAY_LOG(FATAL)

File: src/ray/core_worker/reference_count.cc
Comment (iycheng) [inline src/ray/core_worker/reference_count.cc:769]: Before you merge the PR, let's run the micro benchmark  to make sure there is no regression.

File: src/ray/pubsub/subscriber.cc
Comment (iycheng) [inline src/ray/pubsub/subscriber.cc:79]: No need for std::shared_ptr. SubscriptionCallback should be enough.

You can return nullptr and do checking like ptr.

File: src/ray/pubsub/subscriber.cc
Comment (iycheng) [inline src/ray/pubsub/subscriber.cc:98]: Can we return as a copy?

File: src/ray/pubsub/subscriber.cc
Comment (iycheng) [inline src/ray/pubsub/subscriber.cc:116]: Why do you need this? I see later we have another failure_callback defined in the for loop.

File: src/ray/pubsub/subscriber.cc
Comment (iycheng) [inline src/ray/pubsub/subscriber.cc:329]:
```suggestion
commands_[publisher_id].push(std::move(command));
```

Comment (iycheng) [reply src/ray/pubsub/subscriber.cc:329]: It'll be created automatically if not exists in cpp.

Comment (iycheng) [inline src/ray/pubsub/subscriber.cc:329]: commands_[publisher_id].emplace(std::move(command));

File: src/ray/pubsub/subscriber.cc
Comment (iycheng) [inline src/ray/pubsub/subscriber.cc:227]: Is it used any where?

File: src/ray/pubsub/subscriber.cc
Comment (iycheng) [inline src/ray/pubsub/subscriber.cc:245]: I feel bad about this one. Could you do some effort to avoid explicit unlock? I don't even know why it's locked here. :(

File: src/ray/pubsub/subscriber.cc
Comment (iycheng) [inline src/ray/pubsub/subscriber.cc:256]: the same. :(

File: src/ray/core_worker/core_worker.cc
Comment (rkooo567) [reply src/ray/core_worker/core_worker.cc:2465]: Hmm this is not the scope of this PR, but sure I will change that.

File: src/ray/core_worker/core_worker.cc
Comment (rkooo567) [reply src/ray/core_worker/core_worker.cc:2489]: This is actually the same thing. You mean just RAY_LOG(FATAL) instead of having false here? 

```
#define RAY_CHECK(condition)                                                          \
  (condition)                                                                         \
      ? RAY_IGNORE_EXPR(0)                                                            \
      : ::ray::Voidify() & ::ray::RayLog(__FILE__, __LINE__, ray::RayLogLevel::FATAL) \
                               << " Check failed: " #condition " "
```

File: src/ray/core_worker/reference_count.cc
Comment (rkooo567) [reply src/ray/core_worker/reference_count.cc:769]: Where do you see the performance could be impacted? Is it because we need to lock/unlock more often as we have locks for both subscriber and here?

File: src/ray/pubsub/subscriber.cc
Comment (rkooo567) [reply src/ray/pubsub/subscriber.cc:79]: I actually have a cpp question here. Isn't the callback including captured values? (thus pretty expensive to copy)?

File: src/ray/pubsub/subscriber.cc
Comment (rkooo567) [reply src/ray/pubsub/subscriber.cc:98]: Same question as above.

File: src/ray/pubsub/subscriber.cc
Comment (rkooo567) [reply src/ray/pubsub/subscriber.cc:116]: Will remove.

File: src/ray/pubsub/subscriber.cc
Comment (rkooo567) [reply src/ray/pubsub/subscriber.cc:227]: will remove it is not used.

File: src/ray/pubsub/subscriber.cc
Comment (rkooo567) [reply src/ray/pubsub/subscriber.cc:245]: Actually the reasoning is inside the comment. There are various fields that need to be protected in this method. And we should unlock here otherwise subscribe callback will deadlock if it uses any Subscriber APIs.

Comment (rkooo567) [reply src/ray/pubsub/subscriber.cc:245]: This needs to be protected because this method is called within the client callback, which runs on a different io service. So the method should be locked.

Comment (rkooo567) [reply src/ray/pubsub/subscriber.cc:245]: Also it is okay to unlock for all code block inside (it just calls a callback). Alternative approach is posting them to io service, but this can add pretty big overhead

File: src/ray/pubsub/subscriber.cc
Comment (rkooo567) [reply src/ray/pubsub/subscriber.cc:329]: You mean .push right? commands is <id, std::queue>

**Review by rkooo567 — COMMENTED** [review]
I have several follow up responses!

File: src/ray/pubsub/subscriber.cc
Comment (rkooo567) [reply src/ray/pubsub/subscriber.cc:245]: The third approach is to use reentrantlock lock (but I didn't really like this approach)

Comment (rkooo567) [reply src/ray/pubsub/subscriber.cc:245]: Okay. So there are 4 approaches. Lmk which one you prefer the most; 

- The current (explicit unlock)
- pass io service to subscriber. This can make things very ugly if the client pool doesn't use the same io service as passed io service (so I don't prefer it).
- re-entarant lock
- Lock twice (before this block + after this block) explictly

Comment (iycheng) [reply src/ray/pubsub/subscriber.cc:242]: I think we talked about post the job to io services in core worker (not the subscriber). Did I miss anything? The reason this is bad is that it's hard to understand even with the comment.

File: src/ray/core_worker/core_worker.cc
Comment (iycheng) [reply src/ray/core_worker/core_worker.cc]: Thanks!
It's always good to just update the easy to improve code like this one along with the PR to improve the overall code quality. Otherwise, I think this probably won't be updated forever 😀

File: src/ray/core_worker/core_worker.cc
Comment (iycheng) [reply src/ray/core_worker/core_worker.cc]: Yes, RAY_FATAL instead of RAY_CHECK(false) for the case fail for sure. It's fatal!

File: src/ray/core_worker/reference_count.cc
Comment (iycheng) [reply src/ray/core_worker/reference_count.cc:761]: Whenever when you have lock it might (or might not) impact the performance, and running microbenchmark is the easiest way to make sure it's not causing any trouble.

**Comment by iycheng** [comment]
I think we talked about the way to post it to io worker in core worker for the job to avoid the lock. If you feel good about this could you please create an issue and make sure to remove the lock before you cut off with this work? I'm ok it can be added later.

File: src/ray/pubsub/subscriber.cc
Comment (rkooo567) [reply src/ray/pubsub/subscriber.cc:242]: Let me make an issue and follow up later in a separate PR. I am still not sure if it is a good behavior to make this asynchronous (it is much harder to use /reason for API user perspective whereas the current approach is just a little complex for the module developer, but this sort of unlocking is not too uncommon pattern in our code).

**Comment by rkooo567** [comment]
https://github.com/ray-project/ray/issues/16381 cc @iycheng

File: src/ray/pubsub/subscriber.cc
Comment (iycheng) [reply src/ray/pubsub/subscriber.cc:242]: I'm open to all options but strong against the lock here.

**Comment by iycheng** [comment]
There is shared_ptr not removed.

File: src/ray/pubsub/subscriber.cc
Comment (rkooo567) [reply src/ray/pubsub/subscriber.cc:242]: Let me try understanding what you are against here actually. Is it just that it is hard to understand why we need to unlock here (even with comments)? So using re-entarant lock can be a possible option for you?

Comment (iycheng) [reply src/ray/pubsub/subscriber.cc:242]: With lock here makes things complicated, but I'm ok with giving it a go (not strong opinion).
but we shouldn't have things like:
```
// in one place
{
   lock
}
// in another place
{
  unlock
}
```

Previously, we have 
```
{
   lock_guard(lock)
}
```
I'm fine with this. I remember last time when we have lock and unlock explicitly, I also mentioned that it's hard to understand the code. The thing here is that we can do something to avoid this one, not the case that we have to use lock/unlock explicitly.

This usually is because the thread modeling has some issues which make it hard to implement.

Comment (rkooo567) [reply src/ray/pubsub/subscriber.cc:242]: So, there's no clear alternative that can ensure the same behavior except re-entarant lock? I think making the api caller use io service.post is worse than this because it requires the api caller to think about threading model + implementation detail.

File: src/ray/pubsub/subscriber.cc
Comment (rkooo567) [reply src/ray/pubsub/subscriber.cc:79]: Thanks this is awesome!

File: src/ray/core_worker/core_worker.cc
Comment (iycheng) [inline src/ray/core_worker/core_worker.cc:2536]: `RAY_LOG(FATAL) << "Invalid command has received: " << static_cast<int>(command.command_message_one_of_case())`

File: src/ray/pubsub/subscriber.cc
Comment (iycheng) [inline src/ray/pubsub/subscriber.cc:97]: if you'd like to avoid capture copying, you'll want to do:

`return std::move(maybe_subscription_callback).value()`

**Review by fishbone — COMMENTED** [review]
I feel much better about this version. Thanks for your effort! Btw, some comments. Almost there.

**Review by fishbone — APPROVED** [review]
Nice work thanks!

**Pull request merged**
